satellite/gracefulexit/endpoint_test.go

```diff
@@ -1207,6 +1207,189 @@ func TestFailureNotFoundPieceHashUnverified(t *testing.T) {
+			require.NoError(t, err)
+
+			for {
+				response, err := c.Recv()
+				if unknownMsgSent {
+					require.Error(t, err)
+					break
+				} else {
+					require.NoError(t, err)
+				}
+
+				switch m := response.GetMessage().(type) {
+				case *pb.SatelliteMessage_ExitCompleted:
+					break MessageLoop
+				case *pb.SatelliteMessage_TransferPiece:
+					if notRespondingPiece.IsZero() {
+						notRespondingPiece = m.TransferPiece.OriginalPieceId
+					}
+					if m.TransferPiece.OriginalPieceId == notRespondingPiece {
+						messageCount++
+						unknownMsgSent = true
+						// We send an unknown message because we want to fail the
+						// transfer message request we get from the satellite.
+						// This allows us to keep the conn open but repopulate
+						// the pending queue.
+						err = c.Send(&pb.StorageNodeMessage{})
+						require.NoError(t, c.CloseSend())
+					} else {
+						pieceReader, err := exitingNode.Storage2.Store.Reader(ctx, satellite.ID(), m.TransferPiece.OriginalPieceId)
+						require.NoError(t, err)
+
+						header, err := pieceReader.GetPieceHeader()
+						require.NoError(t, err)
+
+						orderLimit := header.OrderLimit
+						originalPieceHash := &pb.PieceHash{
+							PieceId:   orderLimit.PieceId,
+							Hash:      header.GetHash(),
+							PieceSize: pieceReader.Size(),
+							Timestamp: header.GetCreationTime(),
+							Signature: header.GetSignature(),
+						}
+
+						newPieceHash := &pb.PieceHash{
+							PieceId:   m.TransferPiece.AddressedOrderLimit.Limit.PieceId,
+							Hash:      originalPieceHash.Hash,
+							PieceSize: originalPieceHash.PieceSize,
+							Timestamp: time.Now(),
+						}
+
+						receivingNodeID := nodeFullIDs[m.TransferPiece.AddressedOrderLimit.Limit.StorageNodeId]
+						require.NotNil(t, receivingNodeID)
+						signer := signing.SignerFromFullIdentity(receivingNodeID)
+
+						signedNewPieceHash, err := signing.SignPieceHash(ctx, signer, newPieceHash)
+						require.NoError(t, err)
+
+						success := &pb.StorageNodeMessage{
+							Message: &pb.StorageNodeMessage_Succeeded{
+								Succeeded: &pb.TransferSucceeded{
+									OriginalPieceId:      m.TransferPiece.OriginalPieceId,
+									OriginalPieceHash:    originalPieceHash,
+									OriginalOrderLimit:   &orderLimit,
+									ReplacementPieceHash: signedNewPieceHash,
+								},
+							},
+						}
+						err = c.Send(success)
+						require.NoError(t, err)
+					}
+
+				default:
+					t.FailNow()
+				}
+			}
+		}
+
+		// make sure not responding piece not in queue
+		incompletes, err := satellite.DB.GracefulExit().GetIncomplete(ctx, exitingNode.ID(), 10, 0)
+		require.NoError(t, err)
+
+		for _, inc := range incompletes {
+			pieceID := inc.RootPieceID.Derive(exitingNode.ID(), inc.PieceNum)
+			require.NotEqual(t, notRespondingPiece, pieceID)
+		}
+
+		// check that the exit has completed and we have the correct transferred/failed values
+		progress, err := satellite.DB.GracefulExit().GetProgress(ctx, exitingNode.ID())
```

Comment:
Should we check that `FinishedAt` is not nil too?

Reply:
good idea!